torchao/csrc/cuda/fp6_llm/utils_parallel_dequant.cuh

```diff
@@ -58,10 +62,48 @@ __device__ __forceinline__ uint32_t MultScale(uint32_t PackedFP16Pair, half Scal
     return output;
 }
 
+template<int EXPONENT, int MANTISSA>
+__device__ __forceinline__ uint32_t MultScale(uint32_t PackedBF16Pair, __nv_bfloat16 Scale) {
```

Comment:
Will this help? https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__SINGLE.html#_CPPv46ldexpffi

Reply:
Thanks! This is definitely a less hacky solution :sweat_smile:

Reply:
I was actually thinking about `ldexpf(*BF16_1, BIAS_OFFSET)` (`BF16_1` will be upcast to FP32). Then realize the original code does multiplication in FP16, which is a bit strange to me - not sure if it's faster (I have a feeling the GPU will use FP32 math to calculate it anyway). Anyway, the current change is fine!

Reply:
Random thoughts. `float bias` can be made into a `constexpr` (in fact it can be a BF16 constexpr if you only do BF16 mul later). But I don't think there is a simple way apart from a big if-else clause (`ldexpf` cannot be used in a constexpr I think). Don't mind me, just some random thoughts on useless micro-optimizations :rofl:

Reply:
I'm all for micro-optimizations :). The `bias` is now a `constexpr`:

```cpp
constexpr float power_of_two(int n) {
    return (n == 0) ? 1.0f : 2.0f * power_of_two(n - 1);
}
...
{
    constexpr float BIAS = power_of_two(BIAS_OFFSET);
}
```

I ran the benchmark and I can report that this makes absolutely no difference for performance :smile:

Reply:
Wow, that's an interesting and neat way to force constexpr 😄